eyecite/helpers.py

```diff
@@ -1,21 +1,115 @@
 import re
 from datetime import date
-from typing import List, Optional, Union
+from typing import List, Optional, Tuple, cast
 
 from courts_db import courts
 
 from eyecite.models import (
     CaseCitation,
     CitationBase,
-    FullCaseCitation,
+    ParagraphToken,
     StopWordToken,
     Token,
     TokenOrStr,
     Tokens,
 )
-from eyecite.utils import PAGE_NUMBER_REGEX, strip_punct
+from eyecite.utils import strip_punct
+
+# *** Metadata regexes: ***
+# Regexes used to scan forward or backward from a citation token. NOTE:
+# * Regexes are written in verbose mode. Intentional spaces must be escaped.
+# * In many regexes order matters: options separated by "|" are
+#   tested left to right, so more specific (typically longer) have to come
+#   before less specific.
+
+# Pin cite regex:
+# A pin cite is the part of a citation used to specify a particular section of
+# the referenced document. These may have prefixes, may include paragraph,
+# page, or line references, and may have multiple ranges specified.
+# For some examples see
+# https://github.com/freelawproject/courtlistener/issues/1344#issuecomment-662994948
+PIN_CITE_TOKEN_REGEX = r"""
+    # optional label -- ¶, §, p., pp., pg., n., note, & n.
+    # (longest to shortest):
+    (?:
+        (?:
+            (?:&\ )?note|
+            (?:&\ )?nn?\.?|
+            ¶{1,2}|
+            §{1,2}|
+            \*{1,4}|
+            pg\.?|
+            pp\.?|
+            p\.?|
+        )\ ?  # optional space after label
+    )?
+    (?:
+        # page:paragraph cite, like 123:24-25 or 123:24-124:25:
+        \d+:\d+(?:-\d+(?::\d+)?)?|
+        # page range, like 12 or 12-13:
+        \d+(?:-\d+)?
+    )
+"""
+PIN_CITE_REGEX = rf"""
+    \ ?(
+        (?:at\ )?
+        {PIN_CITE_TOKEN_REGEX},?
+        (?:\ {PIN_CITE_TOKEN_REGEX},?)*
+    )
+"""
+
+
+# Short cite antecedent regex:
+# What case does a short cite refer to? For now, we just capture the previous
+# word optionally followed by a comma. Example: Adarand, 515 U.S. at 241.
+SHORT_CITE_ANTECEDENT_REGEX = r"""
+    ([\w\-.]+),?
+    \   # final space
+"""
+
+
+# Supra cite antecedent regex:
+# What case does a short cite refer to? For now, we just capture the previous
+# word optionally followed by a comma. Example: Adarand, supra.
+# If the previous word is a digit, we capture both that (to store as a volume)
+# and the word before it (to store as antecedent).
+SUPRA_ANTECEDENT_REGEX = r"""
+    (?:
+        ([\w\-.]+),?\ (\d+)|
+        (\d+)|
+        ([\w\-.]+),?
+    )
+    \   # final space
+"""
+
+
+# Post citation regex:
+# Capture metadata after a full cite. For example given the citation "1 U.S. 1"
+# with the following text:
+#   1 U.S. 1, 4-5, 2 S. Ct. 2, 6-7 (4th Cir. 2012) (overruling foo)
+# we want to capture:
+#   pin_cite = 4-5
+#   extra = 2 S. Ct. 2, 6-7
+#   court = 4th Cir.
+#   year = 2012
+#   parenthetical = overruling foo
+POST_CITATION_REGEX = rf"""
```

Comment:
Would it be helpful to add names to these match groups? Not sure if that would make the extraction clearer below (lines 170ff.), instead of doing a bunch of `m[]` calls with the group indices?